File: hacking_with_go/data_leakage.go
package main

import (
    "bufio"
    "fmt"
    "net/http"
    "os"
    "strings"
)

func main() {
    // Prompt the user to enter the target website URL
    var targetURL string
    fmt.Print("Enter your target website URL: ")
    scanner := bufio.NewScanner(os.Stdin)
    scanner.Scan()
    targetURL = scanner.Text()

    // Perform data leakage vulnerability tests
    fmt.Println("Testing for data leakage vulnerabilities on", targetURL)

    // Check for sensitive information in the source code
    sourceCodeLeaks := checkSourceCodeLeaks(targetURL)
    if sourceCodeLeaks {
        fmt.Println("[+] Potential data leakage vulnerability found: Exposed sensitive information in the source code.")
    } else {
        fmt.Println("[-] No source code leakage detected.")
    }

    // Check for sensitive data in HTTP responses
    httpLeaks := checkHTTPLeaks(targetURL)
    if httpLeaks {
        fmt.Println("[+] Potential data leakage vulnerability found: Exposed sensitive data in HTTP responses.")
    } else {
        fmt.Println("[-] No sensitive data found in HTTP responses.")
    }

    // Additional tests can be added as needed

    fmt.Println("Data leakage vulnerability testing complete.")
}

// Function to check for sensitive information in the source code
func checkSourceCodeLeaks(url string) bool {
    response, err := http.Get(url)
    if err != nil {
        fmt.Println("Error:", err)
        return false
    }
    defer response.Body.Close()

    // Read the source code
    scanner := bufio.NewScanner(response.Body)
    for scanner.Scan() {
        line := scanner.Text()
        // Check for sensitive keywords (e.g., password, secret, API key, etc.)
        sensitiveKeywords := []string{"password", "secret", "api_key"} // Add more keywords as needed
        for _, keyword := range sensitiveKeywords {
            if strings.Contains(line, keyword) {
                return true
            }
        }
    }
    return false
}

// Function to check for sensitive data in HTTP responses
func checkHTTPLeaks(url string) bool {
    // Send a GET request to the target URL
    response, err := http.Get(url)
    if err != nil {
        fmt.Println("Error:", err)
        return false
    }
    defer response.Body.Close()

    // Analyze HTTP response headers for sensitive information
    for header, values := range response.Header {
        headerName := strings.ToLower(header)
        // Check for headers that may contain sensitive data (e.g., Set-Cookie)
        if headerName == "set-cookie" {
            for _, value := range values {
                if strings.Contains(value, "secure") || strings.Contains(value, "HttpOnly") {
                    return true
                }
            }
        }
    }

    // Analyze HTTP response body for sensitive information
    scanner := bufio.NewScanner(response.Body)
    for scanner.Scan() {
        line := scanner.Text()
        // Check for sensitive data patterns (e.g., credit card numbers, email addresses)
        if strings.Contains(line, "credit_card") || strings.Contains(line, "email") {
            return true
        }
    }
    return false
}

File: hacking_with_python/data_leakage.py
import requests
from bs4 import BeautifulSoup
from termcolor import colored

def test_data_leakage(target_url):
    print(colored("[*] Testing for Data Leakage Vulnerability on:", "blue"), target_url)
    try:
        response = requests.get(target_url)
        if response.status_code == 200:
            soup = BeautifulSoup(response.content, "html.parser")
            sensitive_keywords = ["password", "credit card", "SSN", "social security number", "login"]
            for keyword in sensitive_keywords:
                if soup.find(text=lambda text: text and keyword.lower() in text.lower()):
                    print(colored("[!] Potential Data Leakage Detected:", "red"), f"Keyword: {keyword}")
            print(colored("[+] Data Leakage Test Complete.", "green"))
        else:
            print(colored("[!] Failed to retrieve website content.", "red"))
    except requests.RequestException as e:
        print(colored("[!] Error:", "red"), e)

def main():
    target_website = input(colored("Enter your target website url: ", "blue"))
    test_data_leakage(target_website)

if __name__ == "__main__":
    main()
